File: 1061. Lexicographically Smallest Equivalent String/sol.java
import java.Math.*;
import java.util.*;


class Solution {

    char[] chars = "abcdefghijklmnopqrstuvwxyz".toCharArray();

    void swap(char from, char to) {
        int fromIndex = from - 'a';
        int toIndex = to - 'a';
        char currentMin = chars[fromIndex];
        char mappedTo = chars[toIndex];
        if(mappedTo < currentMin) {
            for(int i=0; i< chars.length; i++) {
                if(chars[i] == currentMin) {
                    chars[i] = mappedTo;
                }
            }
        }
    }

    public String smallestEquivalentString(String s1, String s2, String baseStr) {
        for(int i = 0; i< s1.length(); i++) {
            swap(s1.charAt(i), s2.charAt(i));
            swap(s2.charAt(i), s1.charAt(i));
        }

        StringBuilder sb = new StringBuilder();
        for(char c : baseStr.toCharArray()) {
            int encoding = c - 'a';
            sb.append(chars[encoding]);
        }

        return sb.toString();
    }
}

File: 1061. Lexicographically Smallest Equivalent String/sol.c
void swap(char from, char to, char* chars) {
    int fromIndex = from - 'a';
    int toIndex = to - 'a';
    char currentMin = chars[fromIndex];
    char mappedTo = chars[toIndex];
    if (mappedTo < currentMin) {
        for (int i = 0; i < 26; i++) {
            if (chars[i] == currentMin) {
                printf("Changing mapping of %c from %c to %c\n", i + 'a',
                       chars[i], mappedTo);
                chars[i] = mappedTo;
            }
        }
    }
}

char* smallestEquivalentString(char* s1, char* s2, char* baseStr) {
    char chars[26] = {'a','b','c','d','e','f','g','h','i','j','k','l','m','n','o','p','q','r','s','t','u','v','w','x','y','z'};

    int index = 0;
    while (s1[index] != '\0') {
        swap(s1[index], s2[index], chars);
        swap(s2[index], s1[index], chars);
        index++;
    }

    index = 0;
    while (baseStr[index] != '\0') {
        int encoding = baseStr[index] - 'a';
        baseStr[index] = chars[encoding];
        index++;
    }

    return baseStr;
}